src/util/netif.cpp

```diff
@@ -0,0 +1,299 @@
+    }
+    return std::nullopt;
+}
+
+#elif defined(__APPLE__)
+
+std::optional<CNetAddr> FromSockAddr(const struct sockaddr* addr)
+{
+    // Check valid length. Note that sa_len is not part of POSIX, and exists on MacOS and some BSDs only, so we can't
+    // do this check in SetSockAddr.
+    if (!(addr->sa_family == AF_INET && addr->sa_len == sizeof(struct sockaddr_in)) &&
+        !(addr->sa_family == AF_INET6 && addr->sa_len == sizeof(struct sockaddr_in6))) {
+        return std::nullopt;
+    }
+
+    // Fill in a CService from the sockaddr, then drop the port part.
+    CService service;
+    if (service.SetSockAddr(addr)) {
+        return (CNetAddr)service;
+    }
+    return std::nullopt;
+}
+
+//! MacOS: Get default gateway from route table. See route(4) for the format.
+std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
+{
+    // net.route.0.inet[6].flags.gateway
+    int mib[] = {CTL_NET, PF_ROUTE, 0, family, NET_RT_FLAGS, RTF_GATEWAY};
+    // The size of the available data is determined by calling sysctl() with oldp=nullptr. See sysctl(3).
+    size_t l = 0;
+    if (sysctl(/*name=*/mib, /*namelen=*/sizeof(mib) / sizeof(int), /*oldp=*/nullptr, /*oldlenp=*/&l, /*newp=*/nullptr, /*newlen=*/0) < 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "Could not get sysctl length of routing table: %s\n", SysErrorString(errno));
+        return std::nullopt;
+    }
+    std::vector<std::byte> buf(l);
+    if (sysctl(/*name=*/mib, /*namelen=*/sizeof(mib) / sizeof(int), /*oldp=*/buf.data(), /*oldlenp=*/&l, /*newp=*/nullptr, /*newlen=*/0) < 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "Could not get sysctl data of routing table: %s\n", SysErrorString(errno));
+        return std::nullopt;
+    }
+    // Iterate over messages (each message is a routing table entry).
+    for (size_t msg_pos = 0; msg_pos < buf.size(); ) {
+        if ((msg_pos + sizeof(rt_msghdr)) > buf.size()) return std::nullopt;
+        const struct rt_msghdr* rt = (const struct rt_msghdr*)(buf.data() + msg_pos);
+        const size_t next_msg_pos = msg_pos + rt->rtm_msglen;
+        if (rt->rtm_msglen < sizeof(rt_msghdr) || next_msg_pos > buf.size()) return std::nullopt;
+        // Iterate over addresses within message, get destination and gateway (if present).
+        // Address data starts after header.
+        size_t sa_pos = msg_pos + sizeof(struct rt_msghdr);
+        std::optional<CNetAddr> dst, gateway;
+        for (int i = 0; i < RTAX_MAX; i++) {
+            if (rt->rtm_addrs & (1 << i)) {
+                // 2 is just sa_len + sa_family, the theoretical minimum size of a socket address.
+                if ((sa_pos + 2) > next_msg_pos) return std::nullopt;
+                const struct sockaddr* sa = (const struct sockaddr*)(buf.data() + sa_pos);
+                if (sa->sa_len < 2 || (sa_pos + sa->sa_len) > next_msg_pos) return std::nullopt;
```

Comment:
9a067b5ac154154ec9008a0e8f2d8f03e994a589: this is triggered on my Intel macOS 14.5 machine. If I comment out the `if` statement it works fine.

Reply:
```
msg_pos=0 sizeof(rt_msghdr)=92 buf.size()=10948
```

Reply:
For IPv4:

```
sa->sa_len=16 sa_pos=92 next_msg_pos=144
sa->sa_len=16 sa_pos=108 next_msg_pos=144
sa->sa_len=0 sa_pos=124 next_msg_pos=144
2024-07-08T12:24:02Z [mapport] [net] portmap: Could not determine IPv4 default gateway
```

For IPv6:

```
sa->sa_len=28 sa_pos=92 next_msg_pos=180
sa->sa_len=28 sa_pos=120 next_msg_pos=180
sa->sa_len=0 sa_pos=148 next_msg_pos=180
2024-07-08T12:24:02Z [mapport] [net] portmap: Could not determine IPv6 default gateway
```

Reply:
Or maybe the following:

```cpp
            if (rt->rtm_addrs & (1 << i)) {
                // 2 is just sa_len + sa_family, the theoretical minimum size of a socket address.
                if ((sa_pos + 2) > next_msg_pos) return std::nullopt;
                const struct sockaddr* sa = (const struct sockaddr*)(buf.data() + sa_pos);
                fprintf(stderr, "i=%d sa->sa_len=%hhu sa_pos=%lu next_msg_pos=%zu\n", i, sa->sa_len, sa_pos, next_msg_pos);
                if (sa->sa_len >= 2 && (sa_pos + sa->sa_len) <= next_msg_pos) {
                    if (i == RTAX_DST) {
                        dst = FromSockAddr(sa);
                    } else if (i == RTAX_GATEWAY) {
                        gateway = FromSockAddr(sa);
                    }
                }
                // Skip to next address.
                sa_pos += sa->sa_len;
```

On macOS `RTAX_DST` is 0 and `RTAX_GATEWAY` is 1. 

```
i=0 sa->sa_len=16 sa_pos=92 next_msg_pos=144
i=1 sa->sa_len=16 sa_pos=108 next_msg_pos=144
i=2 sa->sa_len=0 sa_pos=124 next_msg_pos=144
i=5 sa->sa_len=0 sa_pos=124 next_msg_pos=144
```

As for the i=2 and i=5:

```cpp
/*
 * Index offsets for sockaddr array for alternate internal encoding.
 */
#define RTAX_DST        0       /* destination sockaddr present */
#define RTAX_GATEWAY    1       /* gateway sockaddr present */
#define RTAX_NETMASK    2       /* netmask sockaddr present */
#define RTAX_GENMASK    3       /* cloning mask sockaddr present */
#define RTAX_IFP        4       /* interface name sockaddr present */
#define RTAX_IFA        5       /* interface addr sockaddr present */
#define RTAX_AUTHOR     6       /* sockaddr for author of redirect */
#define RTAX_BRD        7       /* for NEWADDR, broadcast or p-p dest addr */
#define RTAX_MAX        8       /* size of array to allocate */
```

Reply:
hm... `sa_len=0` is really strange because it doesn't know how to advance, there isn't even place for the size! (which it just read in the line above it!) it should at least break out of the loop in that case

Reply:
I'm a bit confused where you want this code to go...

Reply:
yeah no problem, i'll push the change as a commit (after rebase, i guess)

Reply:
It works again with 26436dce049d4215a112a1147b624c0435dbc8a6.

Reply:
Thanks for testing, apperently we over-simplified that. WIl keep it like this.